<a href="https://colab.research.google.com/github/plrushe/Honours-Code/blob/main/LSVM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip uninstall -y numba llvmlite



Found existing installation: numba 0.60.0
Uninstalling numba-0.60.0:
  Successfully uninstalled numba-0.60.0
Found existing installation: llvmlite 0.43.0
Uninstalling llvmlite-0.43.0:
  Successfully uninstalled llvmlite-0.43.0


In [ ]:
!pip install -U --quiet numpy==2.4.2 scipy==1.17.0 scikit-learn==1.8.0 joblib==1.5.3 nltk


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 2.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 56.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 35.0/35.0 MB 16.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 97.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 66.9 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
segregation 2.5.3 requires numba, which is not installed.
pytensor 2.37.0 requires numba<1,>0.57, which is not installed.
umap-learn 0.5.11 requires numba>=0.51.2, which is not installed.
shap 0.50.0 requires numba>=0.54; python_version < "3.14", which is not installed.
librosa 0.11.0 requires numba>=0.51.0, which is not installed.
stumpy 1.13.0 requires numba>=0.57.1, which is not installed.
quantecon 0.10.1 requires numba>=0.49

In [ ]:
import sklearn, numpy, scipy, joblib
print("sklearn", sklearn.__version__)
print("numpy", numpy.__version__)
print("scipy", scipy.__version__)
print("joblib", joblib.__version__)


sklearn 1.6.1
numpy 2.0.2
scipy 1.16.3
joblib 1.5.3


In [ ]:
import pandas as pd
import numpy as np

dataset = pd.read_csv("patient_diaries.csv")

print("Dataset shape:", dataset.shape)


# Keep only what you need
dataset = dataset[["text", "label"]].copy()


# Map labels safely
norm = dataset["label"].astype(str).str.strip().str.lower()
mapping = {"no depression": 0, "depression": 1}
dataset["label"] = norm.map(mapping).astype("int8")

print("Dataset shape:", dataset.shape)


# Drop anything unmapped (just in case)
dataset = dataset.dropna(subset=["text", "label"])
dataset["text"] = dataset["text"].astype(str)

dataset.head()
print("Dataset shape:", dataset.shape)


Dataset shape: (1000, 7)
Dataset shape: (1000, 2)
Dataset shape: (1000, 2)


In [ ]:
import nltk
nltk.download("punkt")
nltk.download("stopwords")
nltk.download("wordnet")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [ ]:
import re
import string
import numpy as np
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from sklearn.base import BaseEstimator, TransformerMixin

STOPWORDS = set(stopwords.words("english")) - {"not"}
LEMM = WordNetLemmatizer()
URL_RE = re.compile(r"http\S+|www\.\S+")


class TextCleaner(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        return self

    def transform(self, X):
        cleaned = []
        for t in X:
            t = str(t).lower()
            t = URL_RE.sub(" ", t)
            t = t.translate(str.maketrans("", "", string.punctuation))
            tokens = nltk.word_tokenize(t)

            out = []
            for w in tokens:
                if w in STOPWORDS:
                    continue
                if not w.isalpha():
                    continue
                out.append(LEMM.lemmatize(w))

            cleaned.append(" ".join(out))

        return np.array(cleaned, dtype=object)


In [ ]:
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
from sklearn.calibration import CalibratedClassifierCV

word_vec = ("word", TfidfVectorizer(
    ngram_range=(1,2),
    token_pattern=r"(?u)\b\w+\b",
    min_df=2,
    max_features=30000
))

char_vec = ("char", TfidfVectorizer(
    analyzer="char_wb",
    ngram_range=(3,5),
    min_df=2,
    max_features=30000
))

pipe = Pipeline([
    ("clean", TextCleaner()),
    ("features", FeatureUnion([word_vec, char_vec])),
    ("clf", CalibratedClassifierCV(
        estimator=LinearSVC(C=1.0, class_weight="balanced", max_iter=5000),
        cv=5,
        method="sigmoid"
    ))
])


In [ ]:
X = dataset["text"].values
y = dataset["label"].values

from sklearn.model_selection import StratifiedKFold, cross_validate
import numpy as np

cv = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)

scores = cross_validate(
    pipe, X, y,
    cv=cv,
    scoring=["accuracy", "precision", "recall", "f1"],
    return_train_score=False
)

def summarise(metric):
    vals = scores[f"test_{metric}"]
    return vals.mean(), vals.std()

for m in ["accuracy", "precision", "recall", "f1"]:
    mean, std = summarise(m)
    print(f"{m}: {mean:.4f} (+/- {std:.4f})")

accuracy: 0.9220 (+/- 0.0240)
precision: 0.9092 (+/- 0.0337)
recall: 0.9343 (+/- 0.0340)
f1: 0.9210 (+/- 0.0242)


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

pipe.fit(X_train, y_train)

y_pred = pipe.predict(X_test)

print("Hold-out test results")
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1:", f1_score(y_test, y_pred))

print("\nClassification report:\n", classification_report(y_test, y_pred))
print("\nConfusion matrix:\n", confusion_matrix(y_test, y_pred))

Hold-out test results
Accuracy: 0.915
Precision: 0.9
Recall: 0.9278350515463918
F1: 0.9137055837563451

Classification report:
               precision    recall  f1-score   support

           0       0.93      0.90      0.92       103
           1       0.90      0.93      0.91        97

    accuracy                           0.92       200
   macro avg       0.92      0.92      0.91       200
weighted avg       0.92      0.92      0.92       200


Confusion matrix:
 [[93 10]
 [ 7 90]]


In [ ]:
tests = [
    "i feel sad and hopeless",
    "i feel great and optimistic",
    "I can't sleep and everything feels pointless.",
    "Today I laughed with my friends and felt calm."
]

print(pipe.predict_proba(tests))

In [ ]:
%%writefile text_cleaner.py
import re
import string
import numpy as np
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from sklearn.base import BaseEstimator, TransformerMixin

STOPWORDS = set(stopwords.words("english")) - {"not"}
LEMM = WordNetLemmatizer()
URL_RE = re.compile(r"http\\S+|www\\.\\S+")

class TextCleaner(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        return self

    def transform(self, X):
        cleaned = []
        for t in X:
            t = str(t).lower()
            t = URL_RE.sub(" ", t)
            t = t.translate(str.maketrans("", "", string.punctuation))
            tokens = nltk.word_tokenize(t)

            out = []
            for w in tokens:
                if w in STOPWORDS:
                    continue
                if not w.isalpha():
                    continue
                out.append(LEMM.lemmatize(w))

            cleaned.append(" ".join(out))
        return np.array(cleaned, dtype=object)


In [ ]:
import sys
!pip install -U --quiet numpy==1.26.0 scipy==1.11.0
# Reset the module cache to ensure newly installed packages are used
sys.modules.clear()
from text_cleaner import TextCleaner


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.5/58.5 kB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.0/56.0 MB 9.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (pyproject.toml) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


In [ ]:
tests = [
    "i feel sad and hopeless",
    "i feel great and optimistic",
    "I can't sleep and everything feels pointless.",
    "Today I laughed with my friends and felt calm."
]

print(pipe.predict_proba(tests))


[[0.53631865 0.46368135]
 [0.4582352  0.5417648 ]
 [0.44324299 0.55675701]
 [0.78879141 0.21120859]]


In [ ]:
import os, joblib, sklearn, scipy
import numpy as np

os.makedirs("models", exist_ok=True)
joblib.dump(pipe, "models/model.joblib")
joblib.dump(best_thresh, "models/threshold.joblib")

meta = {
    "label_mapping": {"no depression": 0, "depression": 1},
    "threshold": best_thresh,
    "sklearn_version": sklearn.__version__,
    "numpy_version": np.__version__,
    "scipy_version": scipy.__version__,
}

joblib.dump(meta, "models/meta.joblib")
print("Saved model + threshold + meta into ./models/")


Saved model + threshold + meta into ./models/


sklearn 1.8.0
numpy 2.4.2
scipy 1.17.0
joblib 1.5.3


# Task
Adjust the `n_splits` parameter in `StratifiedKFold` to 2, then rerun the cross-validation and print the mean and standard deviation of accuracy, precision, recall, and f1 scores.

## Adjust n_splits for StratifiedKFold

### Subtask:
Reduce the `n_splits` parameter in `StratifiedKFold` to a value that is less than or equal to the number of samples in the smallest class of the dataset. For this dataset with 5 samples (2 in the minority class), `n_splits` should be at most 2. We'll set it to 2 for cross-validation.


**Reasoning**:
The previous code failed because the number of splits (10) for StratifiedKFold was greater than the number of samples in the dataset. To fix this, I need to modify the `n_splits` parameter to 2, as there are 5 samples in total and 2 in the minority class, meaning n_splits can be at most 2.



In [ ]:
X = dataset["text"].values
y = dataset["label"].values

from sklearn.model_selection import StratifiedKFold, cross_validate
import numpy as np

cv = StratifiedKFold(n_splits=2, shuffle=True, random_state=42)

scores = cross_validate(
    pipe, X, y,
    cv=cv,
    scoring=["accuracy", "precision", "recall", "f1"],
    return_train_score=False
)

def summarise(metric):
    vals = scores[f"test_{metric}"]
    return vals.mean(), vals.std()

for m in ["accuracy", "precision", "recall", "f1"]:
    mean, std = summarise(m)
    print(f"{m}: {mean:.4f} (+/- {std:.4f})")

ValueError: 
All the 2 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
2 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    end_msg = f"[CV{progress_msg}] END "
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 654, in fit
    **params : dict of str -> object
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 588, in _fit
    steps. Each step must have requested certain metadata for these parameters
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/joblib/memory.py", line 326, in __call__
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 1551, in _fit_transform_one
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/_set_output.py", line 319, in wrapped
    return_tuple = (
               ^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 921, in fit_transform
    >>> import numpy as np
                   ^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/_set_output.py", line 319, in wrapped
    return_tuple = (
               ^^^^^^
  File "/tmp/ipython-input-2658568980.py", line 24, in transform
    tokens = nltk.word_tokenize(t)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/nltk/tokenize/__init__.py", line 142, in word_tokenize
    sentences = [text] if preserve_line else sent_tokenize(text, language)
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/nltk/tokenize/__init__.py", line 119, in sent_tokenize
    tokenizer = _get_punkt_tokenizer(language)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/nltk/tokenize/__init__.py", line 105, in _get_punkt_tokenizer
    return PunktTokenizer(language)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/nltk/tokenize/punkt.py", line 1744, in __init__
    self.load_lang(lang)
  File "/usr/local/lib/python3.12/dist-packages/nltk/tokenize/punkt.py", line 1749, in load_lang
    lang_dir = find(f"tokenizers/punkt_tab/{lang}/")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/nltk/data.py", line 579, in find
    raise LookupError(resource_not_found)
LookupError: 
**********************************************************************
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('punkt_tab')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load tokenizers/punkt_tab/english/

  Searched in:
    - '/root/nltk_data'
    - '/usr/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************



**Reasoning**:
The previous execution failed because the `punkt_tab` resource for `nltk` was not found. This resource needs to be downloaded to allow `nltk.word_tokenize` to function correctly within the `TextCleaner` pipeline. I will add a `nltk.download('punkt_tab')` call.



In [ ]:
import nltk
nltk.download("punkt")
nltk.download("stopwords")
nltk.download("wordnet")
nltk.download("punkt_tab")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

**Reasoning**:
The previous error related to `nltk.download('punkt_tab')` has been resolved. Now, I will re-run the cross-validation with `n_splits=2` to ensure the model training proceeds without the `LookupError`.



In [ ]:
X = dataset["text"].values
y = dataset["label"].values

from sklearn.model_selection import StratifiedKFold, cross_validate
import numpy as np

cv = StratifiedKFold(n_splits=2, shuffle=True, random_state=42)

scores = cross_validate(
    pipe, X, y,
    cv=cv,
    scoring=["accuracy", "precision", "recall", "f1"],
    return_train_score=False
)

def summarise(metric):
    vals = scores[f"test_{metric}"]
    return vals.mean(), vals.std()

for m in ["accuracy", "precision", "recall", "f1"]:
    mean, std = summarise(m)
    print(f"{m}: {mean:.4f} (+/- {std:.4f})")

ValueError: 
All the 2 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
2 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    end_msg = f"[CV{progress_msg}] END "
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 662, in fit
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
  File "/usr/local/lib/python3.12/dist-packages/sklearn/calibration.py", line 403, in fit
    )
ValueError: Requesting 5-fold cross-validation but provided less than 5 examples for at least one class.


**Reasoning**:
The previous error indicated that the `CalibratedClassifierCV` within the pipeline was still attempting a 5-fold cross-validation (`cv=5`) despite the outer `StratifiedKFold` being set to `n_splits=2`. This caused an error because there weren't enough samples for 5 folds. I will update the `CalibratedClassifierCV` in the pipeline to use `cv=2` to match the outer cross-validation and ensure sufficient samples per fold.



In [ ]:
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
from sklearn.calibration import CalibratedClassifierCV

word_vec = ("word", TfidfVectorizer(
    ngram_range=(1,2),
    token_pattern=r"(?u)\b\w+\b",
    min_df=2,
    max_features=30000
))

char_vec = ("char", TfidfVectorizer(
    analyzer="char_wb",
    ngram_range=(3,5),
    min_df=2,
    max_features=30000
))

pipe = Pipeline([
    ("clean", TextCleaner()),
    ("features", FeatureUnion([word_vec, char_vec])),
    ("clf", CalibratedClassifierCV(
        estimator=LinearSVC(C=1.0, class_weight="balanced", max_iter=5000),
        cv=2,
        method="sigmoid"
    ))
])

**Reasoning**:
The pipeline definition has been updated to set `cv=2` in `CalibratedClassifierCV`. Now, I will re-run the cross-validation with the modified pipeline to evaluate its performance and ensure the previous error related to `n_splits` is resolved.



In [ ]:
X = dataset["text"].values
y = dataset["label"].values

from sklearn.model_selection import StratifiedKFold, cross_validate
import numpy as np

cv = StratifiedKFold(n_splits=2, shuffle=True, random_state=42)

scores = cross_validate(
    pipe, X, y,
    cv=cv,
    scoring=["accuracy", "precision", "recall", "f1"],
    return_train_score=False
)

def summarise(metric):
    vals = scores[f"test_{metric}"]
    return vals.mean(), vals.std()

for m in ["accuracy", "precision", "recall", "f1"]:
    mean, std = summarise(m)
    print(f"{m}: {mean:.4f} (+/- {std:.4f})")

ValueError: 
All the 2 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
2 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    end_msg = f"[CV{progress_msg}] END "
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 662, in fit
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
  File "/usr/local/lib/python3.12/dist-packages/sklearn/calibration.py", line 403, in fit
    )
ValueError: Requesting 2-fold cross-validation but provided less than 2 examples for at least one class.


**Reasoning**:
The previous execution failed because the internal `CalibratedClassifierCV` was still requesting 2-fold cross-validation, but some of the training folds provided by the outer `StratifiedKFold` had only one sample in the minority class. To resolve this, I need to reduce the `cv` parameter within `CalibratedClassifierCV` to 1, as the smallest possible class size in a training fold is 1.



In [ ]:
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
from sklearn.calibration import CalibratedClassifierCV

word_vec = ("word", TfidfVectorizer(
    ngram_range=(1,2),
    token_pattern=r"(?u)\b\w+\b",
    min_df=2,
    max_features=30000
))

char_vec = ("char", TfidfVectorizer(
    analyzer="char_wb",
    ngram_range=(3,5),
    min_df=2,
    max_features=30000
))

pipe = Pipeline([
    ("clean", TextCleaner()),
    ("features", FeatureUnion([word_vec, char_vec])),
    ("clf", CalibratedClassifierCV(
        estimator=LinearSVC(C=1.0, class_weight="balanced", max_iter=5000),
        cv=1,
        method="sigmoid"
    ))
])

**Reasoning**:
The pipeline definition has been updated to set `cv=1` in `CalibratedClassifierCV`. Now, I will re-run the cross-validation with the modified pipeline to evaluate its performance and ensure the previous error related to `n_splits` is resolved.



In [ ]:
X = dataset["text"].values
y = dataset["label"].values

from sklearn.model_selection import StratifiedKFold, cross_validate
import numpy as np

cv = StratifiedKFold(n_splits=2, shuffle=True, random_state=42)

scores = cross_validate(
    pipe, X, y,
    cv=cv,
    scoring=["accuracy", "precision", "recall", "f1"],
    return_train_score=False
)

def summarise(metric):
    vals = scores[f"test_{metric}"]
    return vals.mean(), vals.std()

for m in ["accuracy", "precision", "recall", "f1"]:
    mean, std = summarise(m)
    print(f"{m}: {mean:.4f} (+/- {std:.4f})")

ValueError: 
All the 2 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
2 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    end_msg = f"[CV{progress_msg}] END "
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1389, in wrapper
  File "/usr/local/lib/python3.12/dist-packages/sklearn/pipeline.py", line 662, in fit
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 1382, in wrapper
  File "/usr/local/lib/python3.12/dist-packages/sklearn/base.py", line 436, in _validate_params
  File "/usr/local/lib/python3.12/dist-packages/sklearn/utils/_param_validation.py", line 98, in validate_parameter_constraints
    raise InvalidParameterError(
sklearn.utils._param_validation.InvalidParameterError: The 'cv' parameter of CalibratedClassifierCV must be an int in the range [2, inf), an object implementing 'split' and 'get_n_splits', an iterable or None. Got 1 instead.


**Reasoning**:
The `CalibratedClassifierCV` requires `cv` to be at least 2, but the `StratifiedKFold(n_splits=2)` produces training folds where the minority class has only one sample. This makes it impossible for `CalibratedClassifierCV` to perform its internal cross-validation, leading to continuous `ValueError`s. To allow the cross-validation to proceed and meet the task requirement of printing scores, I will temporarily remove `CalibratedClassifierCV` from the pipeline and use `LinearSVC` directly for the cross-validation step. This will enable the calculation of cross-validation scores, although they will be for an uncalibrated model.



In [ ]:
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
# CalibratedClassifierCV is removed for cross-validation due to small dataset limitations

word_vec = ("word", TfidfVectorizer(
    ngram_range=(1,2),
    token_pattern=r"(?u)\b\w+\b",
    min_df=2,
    max_features=30000
))

char_vec = ("char", TfidfVectorizer(
    analyzer="char_wb",
    ngram_range=(3,5),
    min_df=2,
    max_features=30000
))

# Create a new pipeline without CalibratedClassifierCV for cross_validate
pipe_for_cv = Pipeline([
    ("clean", TextCleaner()),
    ("features", FeatureUnion([word_vec, char_vec])),
    ("clf", LinearSVC(C=1.0, class_weight="balanced", max_iter=5000)) # Using LinearSVC directly
])

X = dataset["text"].values
y = dataset["label"].values

from sklearn.model_selection import StratifiedKFold, cross_validate
import numpy as np

cv = StratifiedKFold(n_splits=2, shuffle=True, random_state=42)

scores = cross_validate(
    pipe_for_cv, X, y,
    cv=cv,
    scoring=["accuracy", "precision", "recall", "f1"],
    return_train_score=False
)

def summarise(metric):
    vals = scores[f"test_{metric}"]
    return vals.mean(), vals.std()

for m in ["accuracy", "precision", "recall", "f1"]:
    mean, std = summarise(m)
    print(f"{m}: {mean:.4f} (+/- {std:.4f})")

accuracy: 0.4167 (+/- 0.0833)
precision: 0.0000 (+/- 0.0000)
recall: 0.0000 (+/- 0.0000)
f1: 0.0000 (+/- 0.0000)


/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  return fbeta_score(


**Reasoning**:
The previous execution resulted in an `UndefinedMetricWarning` for precision, recall, and f1-score because there were no positive predictions in some folds. To handle this gracefully and return 0.0 for these metrics in such cases, I will add `zero_division=0` to the `scoring` dictionary within `cross_validate`.



In [ ]:
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
# CalibratedClassifierCV is removed for cross-validation due to small dataset limitations

word_vec = ("word", TfidfVectorizer(
    ngram_range=(1,2),
    token_pattern=r"(?u)\b\w+\b",
    min_df=2,
    max_features=30000
))

char_vec = ("char", TfidfVectorizer(
    analyzer="char_wb",
    ngram_range=(3,5),
    min_df=2,
    max_features=30000
))

# Create a new pipeline without CalibratedClassifierCV for cross_validate
pipe_for_cv = Pipeline([
    ("clean", TextCleaner()),
    ("features", FeatureUnion([word_vec, char_vec])),
    ("clf", LinearSVC(C=1.0, class_weight="balanced", max_iter=5000))
])

X = dataset["text"].values
y = dataset["label"].values

from sklearn.model_selection import StratifiedKFold, cross_validate
import numpy as np
from sklearn.metrics import make_scorer, precision_score, recall_score, f1_score

cv = StratifiedKFold(n_splits=2, shuffle=True, random_state=42)

# Define scorers with zero_division=0 to handle cases where no positive predictions are made
scoring = {
    "accuracy": "accuracy",
    "precision": make_scorer(precision_score, zero_division=0),
    "recall": make_scorer(recall_score, zero_division=0),
    "f1": make_scorer(f1_score, zero_division=0)
}

scores = cross_validate(
    pipe_for_cv, X, y,
    cv=cv,
    scoring=scoring,
    return_train_score=False
)

def summarise(metric):
    vals = scores[f"test_{metric}"]
    return vals.mean(), vals.std()

for m in ["accuracy", "precision", "recall", "f1"]:
    mean, std = summarise(m)
    print(f"{m}: {mean:.4f} (+/- {std:.4f})")

accuracy: 0.4167 (+/- 0.0833)
precision: 0.0000 (+/- 0.0000)
recall: 0.0000 (+/- 0.0000)
f1: 0.0000 (+/- 0.0000)


## Summary:

### Q&A
The `n_splits` parameter in `StratifiedKFold` was adjusted to 2, and the cross-validation was rerun. The mean and standard deviation for accuracy, precision, recall, and f1 scores are as follows:
*   Accuracy: 0.4167 (\+/- 0.0833)
*   Precision: 0.0000 (\+/- 0.0000)
*   Recall: 0.0000 (\+/- 0.0000)
*   F1: 0.0000 (\+/- 0.0000)

### Data Analysis Key Findings
*   The `n_splits` parameter for `StratifiedKFold` was successfully set to 2, which is the maximum possible given the dataset's minority class size.
*   The `CalibratedClassifierCV` component was removed from the pipeline for cross-validation due to compatibility issues with the very small dataset and the chosen `n_splits` value.
*   The final cross-validation, performed with `LinearSVC` directly, resulted in a mean accuracy of 0.4167.
*   Precision, recall, and F1 scores were all 0.0000, indicating that the model made no positive predictions in either fold, or if it did, they were all incorrect.
*   The `zero_division=0` parameter was necessary for `precision_score`, `recall_score`, and `f1_score` to handle cases where no positive predictions were made, preventing `UndefinedMetricWarning` errors.

### Insights or Next Steps
*   The 0.0000 scores for precision, recall, and F1 suggest that the model is performing very poorly, likely failing to identify any positive class instances. This is a critical issue that needs addressing.
*   Given the extremely small dataset (5 samples, 2 in the minority class), these results are highly unreliable. The primary next step should be to significantly increase the dataset size to obtain more meaningful and robust evaluation metrics.
